contracts/src/interfaces/IGateway.sol

```diff
@@ -110,4 +171,56 @@ interface IGateway {
         uint128 destinationFee,
         uint128 amount
     ) external payable;
+
+    // V2
+
+    // Submit a message for verification and dispatch
+    function v2_submit(
+        InboundMessageV2 calldata message,
+        bytes32[] calldata leafProof,
+        Verification.Proof calldata headerProof,
+        bytes32 rewardAddress
+    ) external;
+
+    // Send an XCM with arbitrary assets to Polkadot Asset Hub
+    //
+    // Params:
+    //   * `xcm` (bytes): SCALE-encoded VersionedXcm message
+    //   * `assets` (bytes[]): Array of asset specs, constrained to maximum of eight.
+    //
+    // Supported asset specs:
+    // * ERC20: abi.encode(0, tokenAddress, value)
+    //
+    // On Asset Hub, the assets will be received into the assets holding register.
+    //
+    // The `xcm` should contain the necessary instructions to:
+    // 1. Pay XCM execution fees for `xcm`, either from assets in holding,
+    //    or from the sovereign account of `msg.sender`.
+    // 2. Handle the assets in holding, either depositing them into
+    //    some account, or forwarding them to another destination.
+    //
+    // To incentivize message delivery, some amount of ether must be passed and should
+    // at least cover the total cost of delivery to Polkadot. This ether be sent across
+    // the bridge as WETH, and given to the relayer as compensation and incentivization.
+    //
+    function v2_sendMessage(
+        bytes calldata xcm,
+        bytes[] calldata assets,
+        bytes calldata claimer
+    ) external payable;
+
+    // Register Ethereum-native token on AHP, using `xcmFeeAHP` of `msg.value`
+    // to pay for execution on AHP.
+    function v2_registerToken(address token, uint128 xcmFeeAHP) external payable;
+
+    // Register Ethereum-native token on AHK, using `xcmFeeAHP` and `xcmFeeAHK`
+    // of `msg.value` to pay for execution on AHP and AHK respectively.
+    function v2_registerTokenOnKusama(
```

Comment:
Just a thought on registration while thinking about our Security model. Kusama token registration can be built as a separate contract on top of the gateway.

The implementation of a contract called KusamaRegistration, that can build the registration Transact call onchain, and then submit it to the gateway. 

On BH we would decend origin to that Location so the origin would be `{ parents:2 interior: X2(Ethreum{chain_id:1}, KusamaRegistration)}`. 

This origin would be forwarded to AHK using AliasOrigin, and can be used secure the register extrinsic on the foreignAssets pallet on AHK. This is currently how we secure token registration now on AHP, using the gateway contracts origin (Snowbridge Sovereign). 

Upgrade would be just deploying a new version of the KusamaRegistration contract and then updating the origin which is allowed to call register on AHK. And it would be separate from the Gateway.